### ISSUE TITLE
About creating a quic service, can this service be used either as a server or as a client?

### ISSUE THREAD

**Issue by ldhios** [netty/netty-incubator-codec-quic#614]
My requirement is as follows: Initially, I establish hole punching using the underlying UDP protocol of Netty, and once successful, I upgrade to the QUIC protocol.  I would like to know if the QUIC service created by this library can be used as both a server and a client.  I noticed that the examples provided only contain separate server and client implementations.

I have tried looking at the examples for both the client and server. I attempted to initiate a connection to another server by reusing the server's channel. Unfortunately, I encountered persistent timeout errors.

```
@Slf4j
public class UDPServer {
    private EventLoopGroup group;

    private Config config;
    private Peer peer;
    private Integer port;

    private QuicChannel quicChannel;
    private QuicStreamChannel streamChannel;
    private ChannelFuture channelFuture;

    private static UDPServer instance = null;
    public static UDPServer getInstance() {
        if (instance == null) {
            synchronized (UDPServer.class) {
                if (instance == null) {
                    instance = new UDPServer();
                }
            }
        }
        return instance;
    }
    public void init(Config config, Peer peer) {
        this.port = config.getUdpPort();
        this.config = config;
        this.peer = peer;
    }
    public ChannelFuture run() {

        try {

            Bootstrap bootstrap = new Bootstrap();
            this.group = new NioEventLoopGroup();
            bootstrap.group(this.group)
                    .channel(NioDatagramChannel.class)
                    .option(ChannelOption.SO_BROADCAST, true)
                    .option(ChannelOption.SO_REUSEADDR, true)
                    .handler(new ChannelInitializer<NioDatagramChannel>() {
                        @Override
                        protected void initChannel(NioDatagramChannel nioDatagramChannel) throws Exception {
                            final ChannelPipeline pipeline = nioDatagramChannel.pipeline();
                            pipeline.addLast("encoder", new UDPMessageEncoder());
                            pipeline.addLast("decoder", new UDPMessageDecoder());
                            pipeline.addLast(new UDPServerHandle(config, peer));
                            pipeline.addLast(quicHandler());


                        }
                    }).bind(this.port);

            ChannelFuture future = bootstrap.bind(this.port);
            channelFuture = future.channel().closeFuture();
            channelFuture.sync();


        } catch (InterruptedException e) {
            log.error(e.toString());
        } finally {
            group.shutdownGracefully();
        }
        return channelFuture;

    }
    public ChannelFuture updateToQuic() {
        channelFuture.channel().pipeline().remove("encoder");
        channelFuture.channel().pipeline().remove("decoder");
        channelFuture.channel().pipeline().remove(UDPServerHandle.class);

        return channelFuture;
    }

    private ChannelHandler quicHandler() {

        SelfSignedCertificate certificate = null;
        try {
            certificate = new SelfSignedCertificate();
        } catch (CertificateException e) {
            throw new RuntimeException(e);
        }
        QuicSslContext context = QuicSslContextBuilder
                .forServer(certificate.privateKey(), null, certificate.certificate())
                .applicationProtocols("http/0.9").earlyData(true).build();
        ChannelHandler codec = new QuicServerCodecBuilder().sslContext(context)
                .maxIdleTimeout(5000, TimeUnit.MILLISECONDS)
                // 为我们要处理的流（和数据）的最大数量配置一些限制.
                .initialMaxData(10000000)
                .initialMaxStreamDataBidirectionalLocal(1000000)
                .initialMaxStreamDataBidirectionalRemote(1000000)
                .initialMaxStreamsBidirectional(100)
                .initialMaxStreamsUnidirectional(100)
                .tokenHandler(null) // 可自定义 token.
                .handler(new QuicServerHandler()) // 自定义 QuicServerHandler
                .streamHandler(new ChannelInitializer<QuicStreamChannel>() {
                    @Override
                    protected void initChannel(QuicStreamChannel ch) {
//                        ch.pipeline()
//                                .addLast(new LineBasedFrameDecoder(1024))  // 行分隔符解码器
//                                .addLast(new QuicServerStreamHandler());

                        // 自定义处理器
                        ch.pipeline().addLast(new QuicServerStreamHandler());
                    }
                }).build();
        return codec;


    }

    public void quicConnect(String host, Integer port) {
        InetSocketAddress addr = new InetSocketAddress(host, port);
        QuicChannelBootstrap bootstrap = QuicChannel.newBootstrap(this.channelFuture.channel())
                .streamHandler(new ChannelInboundHandlerAdapter() { // 入栈
            @Override
            public void channelActive(ChannelHandlerContext ctx) {

            }
        }).remoteAddress(addr);
        try {
            quicChannel = bootstrap.connect().get();
            streamChannel = createStream(quicChannel).sync().getNow(); // 发送消息通道
            // 等待服务端监听端口关闭
            streamChannel.closeFuture().sync();
            quicChannel.closeFuture().sync();
        } catch (InterruptedException e) {
            throw new RuntimeException(e);
        } catch (ExecutionException e) {
            throw new RuntimeException(e);
        }

    }


    /**
     * 处理器 QuicClientStreamHandler
     */
    public static Future<QuicStreamChannel> createStream(QuicChannel quicChannel) {

        // BIDIRECTIONAL 双向的
        return quicChannel.createStream(QuicStreamType.BIDIRECTIONAL
                , new QuicClientStreamHandler());
    }

}
```

```
io.netty.channel.ConnectTimeoutException: connection timed out: QuicConnectionAddress{EPHEMERAL}
	at io.netty.incubator.codec.quic.QuicheQuicChannel$QuicChannelUnsafe.lambda$connect$1(QuicheQuicChannel.java:1342)
	at io.netty.util.concurrent.PromiseTask.runTask(PromiseTask.java:98)
	at io.netty.util.concurrent.ScheduledFutureTask.run(ScheduledFutureTask.java:153)
	at io.netty.util.concurrent.AbstractEventExecutor.runTask$$$capture(AbstractEventExecutor.java:174)
	at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute$$$capture(AbstractEventExecutor.java:167)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:470)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:503)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:840)
```

**Comment by ldhios** [comment]
I'll add a description of my scenario, using a quic server to connect to another quic server. thank you

**Comment by normanmaurer** [comment]
Yes sure you can use it both as server and client. Just create two DatagramChannel and bootstrap it.

**Comment by ldhios** [comment]
Whether the Quick protocol allows the client and the server to bind the
same port number. Or can the quic server connect to another quic server? I
have tried many times and have some problems, which has troubled me for a
while.


Rick smith ***@***.***>于2023年11月28日 周二22:24写道：

**Comment by normanmaurer** [comment]
You cant bind to the same port.

**Comment by ldhios** [comment]
@normanmaurer  
Thank you very much for your response. As you mentioned, the QUIC client and server cannot use the same port number. Can a QUIC server connect to another QUIC server?

From the perspective of network protocol selection, it seems that I made a mistake from the beginning. Is it not possible to upgrade to QUIC after successfully performing basic UDP hole punching? If possible, could you provide some suggestions or ideas.

**Comment by normanmaurer** [comment]
@ldhios I have no idea about the upgrade but frankly speaking you would need to somehow detect once such a upgrade happens and then adjust the pipeline. A QUIC server can connect to another quic server by creating another DatagramChannel and add the right codecs to the pipeline (basically bootstrap a client that it uses)

**Comment by ldhios** [comment]
Thank you very much for your response. My current situation is that I have implemented a basic UDP protocol and a QUIC server protocol in my service. Here's what I've done: In the first InboundHandler, I check if the incoming packet is a UDP packet. If it is, I handle it in the first handler. If it is a QUIC packet, I forward it to the next QUIC server handler for processing.
  It is working well now, both the inbound and outbound for UDP and the inbound and outbound for QUIC data are functioning without any problems.

1. Now  my service is bound to port 9999, I have successfully performed UDP hole punching using the basic UDP protocol.
2. Since my application is a P2P application, now that I have successfully performed hole punching, I want to connect to another node's QUIC server. To do this, I need to use the service bound to port 9999 to connect to the successfully punched QUIC server. Regarding my previous suggestion of creating another DatagramChannel to the other server, I apologize for any confusion caused. If you create another DatagramChannel, the port number will typically remain the same 9999
